src/ui/public/vis/update_status.ts

```diff
@@ -18,61 +18,43 @@
  */
 
 import { calculateObjectHash } from './lib/calculate_object_hash';
-
-const Status = {
-  AGGS: 'aggs',
-  DATA: 'data',
-  PARAMS: 'params',
-  RESIZE: 'resize',
-  TIME: 'time',
-  UI_STATE: 'uiState',
-};
-
-// adapted from https://github.com/isaacs/json-stringify-safe/blob/02cfafd45f06d076ac4bf0dd28be6738a07a72f9/stringify.js
-function serializer() {
-  const stack = [];
-  const keys = [];
-
-  const cycleReplacer = function (key, value) {
-    if (stack[0] === value) return '[Circular ~]';
-    return `[Circular ~.${keys.slice(0, stack.indexOf(value)).join('.')}]`;
-  };
-
-  return function (key, value) {
-    if (stack.length > 0) {
-      const thisPos = stack.indexOf(this);
-      ~thisPos ? stack.splice(thisPos + 1) : stack.push(this);
-      ~thisPos ? keys.splice(thisPos, Infinity, key) : keys.push(key);
-      if (~stack.indexOf(value)) value = cycleReplacer.call(this, key, value);
-    }
-    else stack.push(value);
-
-    return value;
-  };
+import { Vis } from './vis';
+
+enum Status {
+  AGGS = 'aggs',
+  DATA = 'data',
+  PARAMS = 'params',
+  RESIZE = 'resize',
+  TIME = 'time',
+  UI_STATE = 'uiState',
 }
 
-function getUpdateStatus(requiresUpdateStatus, obj, param) {
+function getUpdateStatus<T extends Status>(
+  requiresUpdateStatus: T[] = [],
+  obj: any,
+  param: { vis: Vis; visData: any; uiState: any }
+): { [reqStats in T]: boolean } {
```

Comment:
This (maybe a bit weird looking syntax) will actually tell TypeScript, that the return value contains exactly those keys, that has been specified in the array passed to the first parameter (since that's where the `T` generic type is used).

That way TypeScript will actually fail on: `getUpdateStatus([Status.TIME], {}, {}).data`, since it knows the return type is of type `{ time: boolean; }`.